src/tiledb/cloud/vcf/ingestion.py

```diff
@@ -0,0 +1,997 @@
+        manifest_samples = set(manifest_df.sample_name)
+        new_samples = manifest_samples.difference(existing_samples)
+        manifest_df = manifest_df[manifest_df.sample_name.isin(new_samples)]
+        result = manifest_df.vcf_uri.to_list()
+        assert result == sorted(result)
+
+        logger.info("%d samples in the manifest.", len(manifest_samples))
+        logger.info("%d samples already ingested.", len(existing_samples))
+        logger.info("%d new samples to ingest.", len(result))
+        prof.write("count", len(result))
+
+    return result
+
+
+def ingest_manifest_udf(
+    dataset_uri: str,
+    sample_uris: Sequence[str],
+    *,
+    config: Optional[Mapping[str, Any]] = None,
+    id: str = "manifest",
+) -> None:
+    """
+    Ingest sample URIs into the manifest array.
+
+    :param dataset_uri: dataset URI
+    :param sample_uris: sample URIs
+    :param config: config dictionary, defaults to None
+    :param id: profiler event id, defaults to "manifest"
+    """
+
+    setup(config)
+
+    with Profiler(group_uri=dataset_uri, group_member=LOG_ARRAY, id=id):
+        group = tiledb.Group(dataset_uri)
+        manifest_uri = group[MANIFEST_ARRAY].uri
+
+        vfs = tiledb.VFS()
+
+        def file_size(uri: str) -> int:
+            try:
+                return vfs.file_size(uri)
+            except Exception:
+                return 0
+
+        def find_index(vcf: str) -> str:
+            for ext in ["tbi", "csi"]:
+                index = f"{vcf}.{ext}"
+                if vfs.is_file(index):
+                    return index
+            return ""
+
+        def get_sample_name(vcf: str) -> str:
+            # TODO: use this when bcftools works with S3
+            if False:
+                sample = subprocess.run(
+                    ["bcftools", "query", "-l", vcf],
+                    capture_output=True,
+                    text=True,
+                ).stdout.strip()
+
+            else:
+                aws_process = subprocess.Popen(
+                    ["aws", "s3", "cp", "--quiet", vcf_uri, "-"],
+                    stdout=subprocess.PIPE,
+                    stderr=sys.stdout,
+                )
+
+                bcftools_process = subprocess.Popen(
+                    ["bcftools", "query", "-l"],
+                    stdin=aws_process.stdout,
+                    stdout=subprocess.PIPE,
+                    stderr=sys.stdout,
+                    text=True,
+                )
+
+                sample = bcftools_process.communicate()[0].strip()
+                aws_process.kill()
+
+            return sample
+
+        def get_records(vcf: str) -> int:
+            # TODO: remove when bcftools index -n works with S3
+            return 0
+
+            cmd = ["bcftools", "index", "-n", vcf]
+            res = subprocess.run(cmd, capture_output=True, text=True)
+            if res.stderr:
+                print(res.stderr)
+                return 0
```

Comment:
This code won’t execute, so this is not a critical thing, but:

Just to make sure—if there is any stderr output, you want to return 0? If that’s what you want to do, that’s fine, but it might be worth adding a comment explaining why.

Reply:
Yes, the 0 in the manifest will indicate there's a problem with the index that needs to be addressed. Added a comment there to explain.